lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryWrite.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.memory;
+
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.Write;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.checkFromIndexSize;
+import static java.util.Objects.requireNonNull;
+
+public class MemoryWrite
+        implements Write
+{
+    private enum State
+    {
+        READY,
+        ABORTED,
+        FINISHED,
+        CLOSED,
+        /**/;
+    }
+
+    public interface OnStreamClose
+    {
+        void onCommit(Slice data)
+                throws IOException;
+    }
+
+    private final Object lock = new Object();
+    private final OnStreamClose onCommit;
+    @GuardedBy("lock")
+    private ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+    @GuardedBy("lock")
+    private State state = State.READY;
+    private final MemoryOutputStream outputStream;
+
+    public MemoryWrite(Location location, OnStreamClose onCommit)
+    {
+        this.onCommit = requireNonNull(onCommit, "onCommit is null");
+        this.outputStream = new MemoryOutputStream(location);
+    }
+
+    @Override
+    public OutputStream stream()
+    {
+        return outputStream;
+    }
+
+    @Override
+    public void finish()
+    {
+        synchronized (lock) {
+            checkState(state == State.READY, "Cannot finish in state %s", state);
+            state = State.FINISHED;
+        }
+    }
```

Comment:
Shouldn't this method throw if the state == State.ABORTED as well?

Reply:
Same with `abort()` shouldn't that throw if state == State.FINISHED?

Reply:
it does; it throws when state!=READY

Reply:
Ah read this wrong